src/storage/data_table.cpp

```diff
@@ -246,6 +258,251 @@ bool DataTable::Delete(transaction::TransactionContext *const txn, const TupleSl
+  }
+  outfile.flush();
+}
+
+void DataTable::WriteDictionaryMessage(std::ofstream &outfile, int64_t dictionary_id,
+                                       const ArrowVarlenColumn &varlen_col,
+                                       flatbuffers::FlatBufferBuilder *flatbuf_builder) {
+  std::vector<flatbuf::FieldNode> field_nodes;
+  std::vector<flatbuf::Buffer> buffers;
+  uint32_t num_elements = varlen_col.OffsetsLength() - 1;
+  size_t buffer_offset = 0;
+  field_nodes.emplace_back(num_elements, 0);
+
+  // Add a fake validity buffer. This is required by flatbuffer. The dictionary message is
+  // actually a warpped recordbatch message, i.e., the dictionary entries are written
+  // in one RecordBatch. RecordBatch is something requires a validity buffer
+  buffers.emplace_back(buffer_offset, 0);
+
+  AddBufferInfo(&buffer_offset, varlen_col.OffsetsLength() * sizeof(uint64_t), &buffers);
+
+  AddBufferInfo(&buffer_offset, varlen_col.ValuesLength(), &buffers);
+
+  auto record_batch =
+      flatbuf::CreateRecordBatch(*flatbuf_builder, num_elements, flatbuf_builder->CreateVectorOfStructs(field_nodes),
+                                 flatbuf_builder->CreateVectorOfStructs(buffers));
+  auto dictionary_batch = flatbuf::CreateDictionaryBatch(*flatbuf_builder, dictionary_id, record_batch);
+  auto aligned_offset = StorageUtil::PadUpToSize(ARROW_ALIGNMENT, buffer_offset);
+  auto message =
+      flatbuf::CreateMessage(*flatbuf_builder, flatbuf::MetadataVersion_V4, flatbuf::MessageHeader_DictionaryBatch,
+                             dictionary_batch.Union(), aligned_offset);
+  flatbuf_builder->Finish(message);
+  int32_t flatbuf_size = flatbuf_builder->GetSize();
+  auto padded_flatbuf_size = StorageUtil::PadUpToSize(ARROW_ALIGNMENT, flatbuf_size);
+  outfile.write(reinterpret_cast<const char *>(&FLATBUF_CONTINUZATION), sizeof(int32_t));
+  outfile.write(reinterpret_cast<const char *>(&padded_flatbuf_size), sizeof(int32_t));
+  outfile.write(reinterpret_cast<const char *>(flatbuf_builder->GetBufferPointer()), flatbuf_size);
+  if (padded_flatbuf_size != static_cast<uint32_t>(flatbuf_size)) {
+    outfile.write(ALIGNMENT, padded_flatbuf_size - flatbuf_size);
+  }
+
+  WriteDataBlock(outfile, reinterpret_cast<const char *>(varlen_col.Offsets()),
+                 varlen_col.OffsetsLength() * sizeof(uint64_t));
+
+  WriteDataBlock(outfile, reinterpret_cast<const char *>(varlen_col.Values()), varlen_col.ValuesLength());
+
+  outfile.flush();
+}
+
+void DataTable::ExportTable(const std::string &file_name) {
+  flatbuffers::FlatBufferBuilder flatbuf_builder;
+  std::ofstream outfile(file_name, std::ios_base::out | std::ios_base::trunc | std::ios_base::binary);
+  std::unordered_map<col_id_t, int64_t> dictionary_ids;
+  WriteSchemaMessage(outfile, &dictionary_ids, &flatbuf_builder);
+
+  const BlockLayout &layout = accessor_.GetBlockLayout();
+  auto column_ids = layout.AllColumns();
+  common::SpinLatch::ScopedSpinLatch guard(&blocks_latch_);
```

Comment:
This will block all the other concurrent scans and inserts into the table while exporting. A fix for now is to copy the list of blocks (`blocks_`) while holding this lock, and release the lock before doing the actual exporting. This is not optimal but good enough for us here (and better than the current version.)

Reply:
Solved by just copying the list of blocks (as suggested). It seems to be safe because the block will never be deallocated until the data table is destroyed.